src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -643,17 +643,31 @@ internal void InitializeForReflectionSerializer()
             // Even if a resolver has already been specified, we need to root
             // the default resolver to gain access to the default converters.
             DefaultJsonTypeInfoResolver defaultResolver = DefaultJsonTypeInfoResolver.RootDefaultInstance();
-            _typeInfoResolver ??= defaultResolver;
+
+            switch (_typeInfoResolver)
+            {
+                case null:
+                    // Use the default reflection-based resolver if no resolver has been specified.
+                    _typeInfoResolver = defaultResolver;
+                    break;
+
+                case JsonSerializerContext ctx when AppContextSwitchHelper.IsSourceGenReflectionFallbackEnabled:
+                    // .NET 6 compatibility mode: enable fallback to reflection metadata for JsonSerializerContext
+                    _effectiveJsonTypeInfoResolver = JsonTypeInfoResolver.Combine(ctx, defaultResolver);
```

Comment:
this will kill fast path, I'd recommend adding the fallback logic to one of the GetTypeInfo/GetConverter instead

Reply:
(don't feel strongly about perf in the compat only scenario though)

Reply:
Fast path consults `_typeInfoResolver` and not `_effectiveJsonTypeInfoResolver`, see

https://github.com/dotnet/runtime/blob/e90b0071089d655445231c8d4d4b74b5e4fc9030/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs#L537